### PR TITLE
Edit pages

### PR DESCRIPTION
**Pull request by tsop14**
- Adds ability to edit projects and events
- Uses automatic for transforming between model and forms

digital_website

### CODE REVIEW

**Review by marialivia16 — APPROVED** [review]
👍 I'll look into fixing the dates

File: app/controllers/Application.scala [hunk 1]
```diff
@@ -73,29 +73,30 @@ class Application(dynamoClient: AmazonDynamoDB, talksTableName: String, eventsTa
...
-  //        dynamoDbService.updateEvent(id, Event(eventData))
-  //        Redirect(routes.Application.events())
-  //      }
-  //    )
-  //  }
+  def editEventPage(id: String) = Action { implicit request =>
+    val event = dynamoDbService.queryEvents(id)
+    event match {
+      case Some(event) => Ok(views.html.editEvent(id, eventForm.fill(EventFormData(event))))
+      case None => ???
```
Comment (philwills) [inline app/controllers/Application.scala:86]: I'd guess you want to return a `NotFound` in this case.

File: app/controllers/Application.scala [hunk 2]
```diff
@@ -113,7 +114,26 @@ class Application(dynamoClient: AmazonDynamoDB, talksTableName: String, eventsTa
...
+        Redirect(routes.Application.projects())
+      }
+    )
+  }
+
+  def editProjectPage(id: String) = Action { implicit request =>
+    val project = dynamoDbService.queryProjects(id)
+    project match {
+      case Some(project) => Ok(views.html.editProject(id, projectForm.fill(ProjectFormData(project))))
+      case None => ???
```
Comment (philwills) [inline app/controllers/Application.scala:127]: `NotFound` would make sense again here.

**Comment by marialivia16** [comment]
@tsop14 You didn't add the `editProject` view page.

File: app/models/forms.scala [hunk 3]
```diff
@@ -13,13 +14,8 @@ object Forms {
...
     def apply(talk: Talk): TalkFormData = {
-      TalkFormData(
-        title = talk.title,
-        url = talk.url,
-        authors = talk.authors.foldLeft(Seq.empty: Seq[AuthorFormData]) { (seq, author) => seq :+ AuthorFormData(author) },
-        location = talk.location,
-        date = talk.date,
-        thumbnail = talk.thumbnail
+      transform[Talk, TalkFormData](talk,
+        "authors" -> talk.authors.foldLeft(Seq.empty: Seq[AuthorFormData]) { (seq, author) => seq :+ AuthorFormData(author) }
```
Comment (philwills) [inline app/models/forms.scala:18]: I thought you could replace this fold with `talk.authors.map(author => AuthorFormData(author))`, but apparently you need `(talk.authors.map(author => AuthorFormData(author)): Seq[AuthorFormData])`. Not sure why the type inference breaks down here, but I think it's still marginally clearer.

Perhaps @cb372 might know what's going on.

File: app/models/models.scala [hunk 4]
```diff
@@ -43,14 +41,9 @@ object Talk {
...
-      id = someId,
-      title = talkData.title,
-      url = talkData.url,
-      authors = talkData.authors.foldLeft(Seq.empty: Seq[Author]) { (seq, authorFormData) => seq :+ Author(authorFormData) },
-      location = talkData.location,
-      date = talkData.date,
-      thumbnail = talkData.thumbnail
+    transform[TalkFormData, Talk](talkData,
+      "id" -> someId,
+      "authors" -> talkData.authors.foldLeft(Seq.empty: Seq[Author]) { (seq, authorFormData) => seq :+ Author(authorFormData) }
```
Comment (philwills) [inline app/models/models.scala:46]: Same comment about `map` applies here.

File: app/models/models.scala [hunk 5]
```diff
@@ -19,11 +20,8 @@ case class Author(
 
 object Author {
   def apply(authorData: AuthorFormData): Author = {
-    Author(
-      id = UUID.randomUUID().toString,
-      name = authorData.name,
-      url = authorData.url,
-      avatar = authorData.avatar
+    transform[AuthorFormData, Author](authorData,
+      "id" -> UUID.randomUUID().toString
```
Comment (philwills) [inline app/models/models.scala:24]: Am I right in thinking that this means Authors can never be edited and re-saved once written?

**Comment by marialivia16** [comment]
@philwills I'll address your suggestions in my next PR. Thanks!

File: app/controllers/Application.scala (hunk 1 above)
Comment (tsop14) [reply app/controllers/Application.scala:86]: @marialivia16 you happy to update this in next PR?

File: app/controllers/Application.scala (hunk 2 above)
Comment (tsop14) [reply app/controllers/Application.scala:127]: @marialivia16 ditto

**Comment by tsop14** [comment]
@marialivia16 

> @tsop14 You didn't add the editProject view page.

Added in #20

**Comment by marialivia16** [comment]
[Edit and delete endpoints for Talks, Events and Projects](https://trello.com/c/qD6nSmvF/13-edit-and-delete-endpoints-for-talks-events-and-projects)

**Pull request merged**